src/utils/display-observer.js

```diff
@@ -196,19 +371,39 @@ export class DisplayObserver {
       }
       seen.add(target);
       const callbacks = this.targetObserverCallbacks_.get(target);
-      if (!callbacks) {
+      const index = findObserverByIo(this.observers_, io);
+      if (!callbacks || index == -1) {
         continue;
       }
-      let observations = this.targetObservations_.get(target);
-      if (!observations) {
-        observations = emptyObservations(this.observers_.length);
-        this.targetObservations_.set(target, observations);
+      this.setObservation_(target, index, isIntersecting, callbacks);
+    }
+  }
+
+  /**
+   * @param {!Element} target
+   * @param {number} index
+   * @param {?boolean} value
+   * @param {?Array<!ObserverCallbackDef>} callbacks
+   * @private
+   */
+  setObservation_(target, index, value, callbacks) {
+    let observations = this.targetObservations_.get(target);
+    if (!observations) {
```

Comment:
I believe this only happens during `observe(target, cb)`, right? If so, we could just move that there and simplify `setObservation_`.

Reply:
I think so. But this still feels a bit safer since the `map.get()` by type returns a `X|undefined` value.